pkg/scheduler/framework/plugins/interpodaffinity/plugin.go

```diff
@@ -132,3 +134,102 @@ func GetNamespaceLabelsSnapshot(logger klog.Logger, ns string, nsLister listersv
+				"pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+			return framework.QueueSkip, nil
+		}
+		logger.V(5).Info("a scheduled pod was updated and it matches the target pod's affinity or doesn't match the target pod's anti-affinity, and the pod may be schedulable now",
+			"pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+		return framework.Queue, nil
+	}
+
+	// Pod is added. Return Queue when the added pod matching the target pod's affinity.
+	if modifiedPod != nil {
+		if podMatchesAllAffinityTerms(terms, modifiedPod) {
+			logger.V(5).Info("a scheduled pod was added and it matches the target pod's affinity",
+				"pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+			return framework.Queue, nil
+		}
+		logger.V(5).Info("a scheduled pod was added and it doesn't match the target pod's affinity",
+			"pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+		return framework.QueueSkip, nil
+	}
+
+	// Pod is deleted. Return Queue when the deleted pod matching the target pod's anti-affinity.
+	for _, t := range antiTerms {
+		if t.Matches(originalPod, nil) {
+			logger.V(5).Info("a scheduled pod was deteled but it matches the target pod's anti-affinity",
+				"pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+			return framework.QueueSkip, nil
+		}
+	}
+	logger.V(5).Info("a scheduled pod was deleted but it doesn't match the target pod's anti-affinity, and the pod may be schedulable now",
+		"pod", klog.KObj(pod), "modifiedPod", klog.KObj(modifiedPod))
+	return framework.Queue, nil
+}
+
+func (pl *InterPodAffinity) isSchedulableAfterNodeChange(logger klog.Logger, pod *v1.Pod, oldObj, newObj interface{}) (framework.QueueingHint, error) {
+	originalNode, modifiedNode, err := util.As[*v1.Node](oldObj, newObj)
+	if err != nil {
+		return framework.Queue, err
+	}
+
+	terms, err := framework.GetAffinityTerms(pod, framework.GetPodAffinityTerms(pod.Spec.Affinity))
+	if err != nil {
+		return framework.Queue, err
+	}
+
+	// Return Queue after a new node was added.
+	if originalNode == nil {
+		for _, term := range terms {
+			if _, ok := modifiedNode.Labels[term.TopologyKey]; !ok {
+				logger.V(5).Info("a new node without matched topologyKey was added", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+				return framework.QueueSkip, nil
+			}
+		}
+		logger.V(5).Info("the pod may be schedulable now because a new node was added", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+		return framework.Queue, err
+	}
+
+	// Return Queue after node topologyKey changes.
+	for _, term := range terms {
+		if modifiedNode.Labels[term.TopologyKey] != originalNode.Labels[term.TopologyKey] {
+			logger.V(5).Info("the pod may be schedulable now after the node topologyKey changed", "pod", klog.KObj(pod), "node", klog.KObj(modifiedNode))
+			return framework.Queue, err
+		}
+	}
```

Comment:
This implementation is correct, _**ideally**_... Similar to [what we did for TopologySpread](https://github.com/kubernetes/kubernetes/blob/48633e879c31d62f91f76a2fd46a8d76d918c476/pkg/scheduler/framework/plugins/podtopologyspread/plugin.go#L258-L261), we have to loosen the check for Node/Update event.

We have to change here to return Queue when `modifiedNode` has any of TopologyKey of affinity, which is the same we do for Node/Add events. So, I guess we end up not distinguishing Node/Add and Node/Update events for now.